Fix: Support embedding packaged submodules in C applications

Re1nyyy opened the pull request:
This PR addresses a critical issue where the use of `cython --embed --embed-modules` with module names containing dots (e.g., pkg.module) leads to C compilation failure due to the generation of invalid C symbols("."). This PR partly solved issue #7365.

Those changes ensures that all generated C symbols derived from Python module names are properly sanitized by replacing dots (.) with underscores (_):

1.  **Changes in Cython/Compiler/ModuleNode.py**: Updated the compiler logic to simplify data passing by removing the redundant C symbol list, ensuring only the full Python module names are passed.
2.  **Changes in Cython/Utility/Embed.c**: Ensured the C code template utilizes the module's base name (e.g., extracting `module` from `pkg.module`) to generate the correct C symbol **`PyInit_module`** for both function declaration and the `PyImport_AppendInittab` call.

Related to:#7365

Here is a example of full process:

### Preparation
1.File Structure
```
cython/
├── main.pyx
└── pkg/
    └── module.pyx
    └── __init__.pyx
```
2.Source Code
#pkg/module.pyx :
```
def hello():
    print("Hello from pkg.module!")
```
#main.pyx :
```
import pkg.module
print("Starting embedded Cython app...")
pkg.module.hello()
print("App finished.")
```
#pkg/__init__.pyx
```
__path__ = []
```
### Step One: Generate C Code
```
1.cython --module-name pkg pkg/__init__.pyx #generate pkg/__init__.c
2.cython --module-name pkg.module pkg/module.pyx #generate pkg/module.c
3.cython --embed --embed-modules=pkg,pkg.module main.pyx #generate main.c
```
### Step Two: Compile and Link
```
1.gcc -c pkg/__init__.c -o pkg_init.o $(python3-config --embed --cflags)
2.gcc -c pkg/module.c -o pkg_module.o $(python3-config --embed --cflags)
3.gcc -c main.c -o main.o $(python3-config --embed --cflags)
4.gcc -o my_embedded_app main.o pkg_init.o pkg_module.o $(python3-config --embed --ldflags)
```
### Step Three: Run and Verify

If linking is successful, run the program.
```
./my_embedded_app
```
Expected Result:
```
Starting embedded Cython app...
Hello from pkg.module!
App finished.
```

scoder commented on Cython/Compiler/ModuleNode.py:
```diff
@@ -3555,14 +3555,19 @@ def generate_main_method(self, env, code):
         else:
             wmain = Options.embed
+        p_embed_modules = (env.module_name,) + tuple(Options.embed_modules)
+        c_embed_modules = (env.module_name,) + tuple(mname.replace('.', '_') for mname in Options.embed_modules)
```
This can be simplified using list unpacking. And I would prefer a list here rather than a tuple because we are dealing with a homogeneous sequence, not a struct-like collection of different values.
```suggestion
        p_embed_modules = [env.module_name, *Options.embed_modules]
        c_embed_modules = [env.module_name, *(mname.replace('.', '_') for mname in Options.embed_modules)]
```
However, this isn't quite how it works. I think we need the original package information in order to generate correct C code.

scoder commented on Cython/Utility/Embed.c:
```diff
@@ -5,8 +5,8 @@
-    __Pyx_PyMODINIT_FUNC PyInit_{{mname}}(void) CYTHON_SMALL_CODE; /*proto*/
+{{for c_mname in c_embed_modules[1:]}} 
+    __Pyx_PyMODINIT_FUNC PyInit_{{c_mname}}(void) CYTHON_SMALL_CODE; /*proto*/
 {{endfor}}
```
The `PyInit_` C function names do not consider the package. They are only based on the module name.
Thus, `module_name.rpartition('.')[-1]` should give you the module name part of the `PyInit_` function name.

Re1nyyy replied:
> The `PyInit_` C function names do not consider the package. They are only based on the module name. Thus, `module_name.rpartition('.')[-1]` should give you the module name part of the `PyInit_` function name.

OK , I will try it later.Thanks for your guidance!

da-woods commented:
Something we should sort out: "embedding" in general is pretty poorly tested and there's definitely some specific cases that we want to test (this one, multiple modules in general, these some multiprocessing stuff that's broken and will need testing). Currently it's only the simple embedding demo that's tested.

I think that should probably be done separately, but when that's set up then this PR definitely needs a test.

scoder commented:
> I have found it could work through just little changes.

Cool. Now we need to make sure that it a) really works and b) keeps working, by adding a test for it. It's too bad that we didn't add a test for this feature in https://github.com/cython/cython/pull/4978 right away, so I wrote one in https://github.com/cython/cython/pull/7371. Once that's merged, you can prove that your PR works by merging the master branch with that test and then extend it to make use of packages.

Re1nyyy commented:
> > I have found it could work through just little changes.
> 
> Cool. Now we need to make sure that it a) really works and b) keeps working, by adding a test for it. It's too bad that we didn't add a test for this feature in #4978 right away, so I wrote one in #7371. Once that's merged, you can prove that your PR works by merging the master branch with that test and then extend it to make use of packages.

OK, I will wait until #7371 is merged and then test and extend it.

Re1nyyy commented:
I think the work towards supporting Unicode should wait until this PR is merged.

scoder commented:
https://github.com/cython/cython/pull/7371 is merged, so this can be tested.

Re1nyyy commented:
@scoder 

Hi,

I've implemented and added the regression test for package structure support (mypkg/submod.pyx) into `embed_modules.srctree`.

However, I'm facing persistent issues running specific tests locally. Even after upgrading my development environment from Ubuntu 20.04 to Ubuntu 22.04 (GDB 12.1, Python 3.10), I am still unable to run the target test using path or tag filtering.

Attempts:
1. `python runtests.py -v tests/build/embed_modules.srctree` results in `Ran 0 tests`.
2. `python runtests.py -v --tag embed` results in `unrecognized arguments: --tag`. 
3. `python3 runtests.py -vv --no-file --no-doctest --no-unit` runs successfully but does not show any output related to the 'embed' tests or my regression test.

I don't think this is due to my OS version, as 22.04 is modern. Could you give me some help?Thanks

Re1nyyy commented:
@scoder 

My code passed `embed_modules.srctree` test. Then I have tried hard to extend the `embed_modules.srctree` test for days, but I don't succeed. I have extended the test like this(one version, didn't passed):

```
CYTHONIZE -j2 pkg/module.pyx
CYTHON --embed --embed-modules pkg.module main_pkg.pyx 
PYTHON build.py main_pkg main_pkg.c pkg/module.c
PYTHON run_pkg.py main_pkg 
################# main_pkg.pyx ####################

import pkg.module
assert __name__ == '__main__'

message = pkg.module.get_message()
print(message)
print("DONE_PKG")


################# pkg/module.pyx ####################

def get_message():
    return "SUCCESS: Embedded pkg.module ran correctly."

################# pkg/__init__.py ####################



################### run_pkg.py ########################

import os
import sys
from subprocess import run
from sysconfig import get_config_var

EXE_EXT = get_config_var('EXE')
PYLIB_DIR = get_config_var("installed_platbase")

PATH = os.environ.get('PATH', '')
sep = os.pathsep
os.environ.update(PATH=f".{sep}{PYLIB_DIR}{sep}{PATH}")

result = run([os.path.abspath(sys.argv[1] + EXE_EXT)], capture_output=True)

assert b"SUCCESS: Embedded pkg.module ran correctly." in result.stdout, result.stdout
assert b"DONE_PKG" in result.stdout, result.stdout
```

`embed.c`and `ModuleNode.py` didn't changed since last commit.(I tried to edit them but this makes things even more complicated, so I gave up.) During the process of extending `embed_modules.srctree` test, I faced with such bugs like :

1. AssertionError: b''
2. ModuleNotFoundError: No module named 'pkg'
3. ModuleNotFoundError: No module named 'pkg.submodule'

Sincerely, since I'm not really familiar with Cython and especially runtest.py, I asked AI for help. But it still didn't work. I changed ways of import and the ways of  define `pkg` and `pkg/module`, but the bugs is still switching from those three.

During debugging, sometimes modules like pkg and pkg/module are registered correctly, but the system just can't find them when I try to import. This might be what's causing these three bugs. 

I've confirmed I can handle the problem in Issue #7365 and generate valid C code, but I can't get my extended runtest to pass yet. So I feel upset now.

If you have time, could you give me some insight or lead me a correct direction? I think maybe I headed for the wrong direction. I will appreciate a lot if you can do me a favor.
